dap_prinz_green_jobs/pipeline/green_measures/occupations/soc_map.py

```diff
@@ -0,0 +1,378 @@
+
+    :param sim_threshold: The similarity threshold for outputting the most similar SOC match.
+    :type sim_threshold: float
+
+    :param top_n_sim_threshold: The similarity threshold for a match being added to a group of SOC matches.
+    :type top_n_sim_threshold: float
+
+    :param minimum_n: The minimum size of a group of SOC matches.
+    :type minimum_n: int
+
+    :param minimum_prop: If a group of SOC matches have a high proportion (>= minimum_prop) of the same SOC being matched, then use this SOC.
+    :type minimum_prop: float
+
+    ----------
+    Methods
+    ----------
+
+    load_process_soc_data():
+        Load the SOC data
+    unique_soc_job_titles(jobtitle_soc_data):
+        Convert the SOC data into a dict where each key is a job title and the value is the SOC code
+        embed_texts(texts):
+                Get sentence embeddings for a list of input texts
+        load(save_embeds=False):
+                Load everything to use this class, calculate SOC embeddings if they weren't inputted, save embeddings if desired
+        find_most_similar_matches(job_titles, job_title_embeddings):
+                Using the inputted job title embeddings and the SOC embeddings, find the full information about the most similar SOC job titles
+        find_most_likely_soc(match_row):
+                For the full match information for one job title, find the most likely SOC (via top match, or group of top matches)
+        get_soc(job_titles, additional_info=False):
+                (main function) For inputted job titles, output the best SOC match, add extra information about matches using the additional_info argument
+
+        ----------
+    Usage
+    ----------
+        soc_mapper = SOCMapper()
+        soc_mapper.load()
+        matches = soc_mapper.get_soc(job_titles=["data scientist", "Assistant nurse", "Senior financial consultant - London"])
+
+    """
+
+    def __init__(
+        self,
+        local: bool = True,
+        embeddings_output_dir: str = "outputs/data/green_occupations/soc_matching/",
+        batch_size: int = 500,
+        match_top_n: int = 10,
+        sim_threshold: float = 0.67,
+        top_n_sim_threshold: float = 0.5,
+        minimum_n: int = 3,
+        minimum_prop: float = 0.5,
+    ):
+        self.local = local
+        self.embeddings_output_dir = embeddings_output_dir
+        self.batch_size = batch_size
+        self.match_top_n = match_top_n
+        self.sim_threshold = sim_threshold
+        self.top_n_sim_threshold = top_n_sim_threshold
+        self.minimum_n = minimum_n
+        self.minimum_prop = minimum_prop
+
+    def load_process_soc_data(self):
+        """
+        Load the job titles to SOC codes dataset as found on the ONS website.
+        A small amount of processing.
+        """
+
+        jobtitle_soc_data = load_job_title_soc()
+        jobtitle_soc_data = jobtitle_soc_data[jobtitle_soc_data["SOC_2020"] != "}}}}"]
+
+        # TO DO: Not sure this is 1:1
```

Comment:
hmmmm wonder if we could ask adeola or someone who's already spent time doing this before

Reply:
we should - but also I checked and it isn't 1:1 - so this def needs revisiting!

Reply:
(it's mostly 1:1 though)